File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head || !head->next) return nullptr;
        ListNode* slow = head;
        ListNode* fast = head->next;
        bool hasCycle = 0;
        while (fast && fast->next) {
            if (slow == fast) {
                hasCycle = 1;
                break;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        if (!hasCycle) return nullptr;
        unordered_map<ListNode*, int> set;
        slow = head;
        while (slow) {
            if (set.count(slow)) return slow;
            else {
                set.insert({slow, slow->val});
            }
            slow = slow->next;
        }
        return nullptr;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None: return None
        slow = head
        fast = head.next
        cycle = False
        while fast and fast.next:
            if slow == fast: 
                cycle = True
                break
            slow = slow.next
            fast = fast.next.next
        if not cycle: return None
        slow = head
        sets = dict()
        while slow:
            if slow in sets : return slow
            else: sets[slow] = slow.val
            slow = slow.next
        return None